adr/0002-scripting-engine.md

```diff
@@ -0,0 +1,170 @@
+
+```yaml
+- description: SonarQube UI Status Check
+  maxRetries: 30
+  script: |
+    r := fetch('https://sonarqube.uds.dev/api/system/status').json()
+    return r['status'] == 'UP'
+```
+
+Risor has a bunch of built-in modules for DevOps use cases and is totally
+pluggable for our own implementation. These modules in particular come
+to mind as potentially useful for delivery:
+
+- [`aws`](https://risor.io/docs/modules/aws)
+- [`vault`](https://risor.io/docs/modules/vault)
+- [`kubernetes`](https://risor.io/docs/modules/kubernetes)
+- [`pgx`](https://risor.io/docs/modules/pgx)
+
+**Note:** `starlet` provides [a number of builtins](https://github.com/1set/starlet/tree/master/lib)
+like `http` and `csv`, which is comparable to those provided by `risor`. The
+big difference is that `starlet` does not aim to support DevOps use cases
+directly, so it would likely never include things like `vault` or `aws` integrations.
+
+Risor can be embedded into `maru` as a library, which means that all
+script execution would happen natively in the Go runtime. This has
+huge advantages for both portability and security. As with the
+[vendoring tools](#vendor-tools) approach, we can continue to ship a
+single binary with minimal-to-zero external dependencies, but with
+the additional advantage of not having to rely on a host-specific shell.
+
+### Alternatives
+
+#### YAML-based DSL
+
+For the specific case on network status checks, we could add support
+for `jsonpath`-based `condition` checks (and align with existing
+`wait.cluster.condition`).
+
+```yaml
+- description: SonarQube UI Status Check
+  maxRetries: 30
+  wait:
+    network:
+      protocol: https
+      address: sonarqube.uds.dev/api/system/status
+      code: 200
+      condition: '{.body.status}'='UP'
+```
+
+Getting the API for `wait.network.condition` would be challenging,
+though. Do we assume the response body is always JSON? If not, what
+sort of expressions would be available for HTML or text responses?
+
+#### Vendor Tools
+
+Note that the original example depended on `curl`. As a result, this
+is not necessarily portable. It may work on WSL, but probably not
+vanilla Powershell. Though it is widely available, most Linux
+distributions also do not ship with `curl`.
+
+```yaml
+tasks:
+  - description: SonarQube UI Status Check
+    maxRetries: 30
+    cmd: |
+      STATUS=$(./uds zarf tools curl -s 'https://sonarqube.uds.dev/api/system/status' | ./uds zarf tools yq '.status')
+      echo "SonarQube system status: ${STATUS}"
+      if [ $STATUS != "UP" ]; then
+        sleep 10
+        exit 1
+      fi
+```
+
+Speaking of Powershell, we would need to maintain lots of Unix command
+rewrites for anything to be reliable. In the above example:
+
+- `sleep 10` -> `Start-Sleep -Seconds 10`[^2]
+- `curl ...` -> probably ok for the most part, but is actually aliased
+  to `Invoke-WebRequest`[^3]
+
+## Decision
+
+
+
+## Consequences
+
```

Comment:
Assuming we go with risor i think it will be good to note that this doesnt actually limit the ability of using the other solutions that could be used today and as such is backwards compatible. It does however give us and our user much more flexibility in defining complex logic in the tasks and creating better, more reusable workflows that are more portable and dont rely on system specific tooling.